{"$graph": [{"class": "CommandLineTool", "hints": [{"dockerPull": "kerstenbreuer/msa_group_compare:2", "class": "DockerRequirement"}, {"coresMin": 1, "ramMin": 1000, "tmpdirMin": 400, "class": "ResourceRequirement"}], "baseCommand": ["Rscript", "--vanilla", "/src/calc_msa_and_dist_mat.R"], "inputs": [{"type": {"type": "enum", "symbols": ["#calc_msa_and_dist_mat.cwl/distance_method/identity", "#calc_msa_and_dist_mat.cwl/distance_method/similarity"]}, "inputBinding": {"position": 16}, "id": "#calc_msa_and_dist_mat.cwl/distance_method"}, {"type": "File", "inputBinding": {"position": 10}, "id": "#calc_msa_and_dist_mat.cwl/fasta_1"}, {"type": "File", "inputBinding": {"position": 11}, "id": "#calc_msa_and_dist_mat.cwl/fasta_2"}, {"type": "string", "inputBinding": {"position": 12}, "id": "#calc_msa_and_dist_mat.cwl/group_name_1"}, {"type": "string", "inputBinding": {"position": 13}, "id": "#calc_msa_and_dist_mat.cwl/group_name_2"}, {"type": {"type": "enum", "symbols": ["#calc_msa_and_dist_mat.cwl/msa_method/ClustalW", "#calc_msa_and_dist_mat.cwl/msa_method/ClustalOmega", "#calc_msa_and_dist_mat.cwl/msa_method/Muscle"]}, "inputBinding": {"position": 14}, "id": "#calc_msa_and_dist_mat.cwl/msa_method"}, {"type": {"type": "enum", "symbols": ["#calc_msa_and_dist_mat.cwl/seq_type/protein", "#calc_msa_and_dist_mat.cwl/seq_type/dna", "#calc_msa_and_dist_mat.cwl/seq_type/rna"]}, "inputBinding": {"position": 15}, "id": "#calc_msa_and_dist_mat.cwl/seq_type"}], "outputs": [{"type": "File", "outputBinding": {"glob": "*.RDS"}, "id": "#calc_msa_and_dist_mat.cwl/dist_rds"}, {"type": "File", "outputBinding": {"glob": "*.tsv"}, "id": "#calc_msa_and_dist_mat.cwl/dist_tsv"}], "id": "#calc_msa_and_dist_mat.cwl"}, {"class": "CommandLineTool", "requirements": [{"class": "InlineJavascriptRequirement"}], "hints": [{"dockerPull": "kerstenbreuer/msa_group_compare:2", "class": "DockerRequirement"}, {"coresMin": 1, "ramMin": 1000, "tmpdirMin": 400, "class": "ResourceRequirement"}], "baseCommand": ["Rscript", "--vanilla", "/src/generate_tree_and_visualize.R"], "arguments": [{"valueFrom": "${\n  if ( inputs.color_by_group ){\n    return(\"TRUE\")\n  }\n  else {\n    return(\"FALSE\")\n  }\n}\n", "position": 11}], "inputs": [{"type": "boolean", "id": "#generate_tree_and_visualize.cwl/color_by_group"}, {"type": "File", "inputBinding": {"position": 10}, "id": "#generate_tree_and_visualize.cwl/dist_rds"}], "outputs": [{"type": "File", "outputBinding": {"glob": "*.png"}, "id": "#generate_tree_and_visualize.cwl/tree_plot"}], "id": "#generate_tree_and_visualize.cwl"}, {"class": "Workflow", "doc": "This workflow compares two groups of sequences by running two steps: \\n\n  1. a MSA aligment to calculate a distance matrix \\n\n  2. generate and plot a phylogenetic tree annotated by group labels \\n\n\\n\nFor more information please see: \\n\nhttps://w3id.org/cwl/view/git/5227e2c45be641c6675c0a24a8b1f72d712fac40/CWL/workflows/msa_group_compare.cwl\n", "inputs": [{"type": "boolean", "default": true, "doc": "Whether to color the phylogenetic tree by group.\n", "id": "#main/color_by_group"}, {"type": {"type": "enum", "symbols": ["#main/distance_method/identity", "#main/distance_method/similarity"]}, "default": "identity", "doc": "Which distance measure to use. Either \"similarity\" or \"identity\".\n", "id": "#main/distance_method"}, {"type": "File", "doc": "Protein / DNA / RNA sequences of first group in fasta format.\n", "id": "#main/fasta_1"}, {"type": "File", "doc": "Protein / DNA / RNA sequences of second group in fasta format.\n", "id": "#main/fasta_2"}, {"type": "string", "default": "group1", "doc": "Name of first group.\n", "id": "#main/group_name_1"}, {"type": "string", "default": "group2", "doc": "Name of second group.\n", "id": "#main/group_name_2"}, {"type": {"type": "enum", "symbols": ["#main/msa_method/ClustalW", "#main/msa_method/ClustalOmega", "#main/msa_method/Muscle"]}, "default": "ClustalW", "doc": "Which multiple sequence alignment algorithm to use. \nOne of \"ClustalW\", \"ClustalOmega\", or \"Muscle\". \nFor more information see: https://www.ebi.ac.uk/Tools/msa/\n", "id": "#main/msa_method"}, {"type": {"type": "enum", "symbols": ["#main/seq_type/protein", "#main/seq_type/dna", "#main/seq_type/rna"]}, "default": "protein", "doc": "The type of sequence. One of \"protein\", \"dna\", or \"rna\".\n", "id": "#main/seq_type"}], "steps": [{"run": "#calc_msa_and_dist_mat.cwl", "in": [{"source": "#main/distance_method", "id": "#main/calc_msa_and_dist_mat/distance_method"}, {"source": "#main/fasta_1", "id": "#main/calc_msa_and_dist_mat/fasta_1"}, {"source": "#main/fasta_2", "id": "#main/calc_msa_and_dist_mat/fasta_2"}, {"source": "#main/group_name_1", "id": "#main/calc_msa_and_dist_mat/group_name_1"}, {"source": "#main/group_name_2", "id": "#main/calc_msa_and_dist_mat/group_name_2"}, {"source": "#main/msa_method", "id": "#main/calc_msa_and_dist_mat/msa_method"}, {"source": "#main/seq_type", "id": "#main/calc_msa_and_dist_mat/seq_type"}], "out": ["#main/calc_msa_and_dist_mat/dist_tsv", "#main/calc_msa_and_dist_mat/dist_rds"], "id": "#main/calc_msa_and_dist_mat"}, {"run": "#generate_tree_and_visualize.cwl", "in": [{"source": "#main/color_by_group", "id": "#main/generate_tree_and_visualize/color_by_group"}, {"source": "#main/calc_msa_and_dist_mat/dist_rds", "id": "#main/generate_tree_and_visualize/dist_rds"}], "out": ["#main/generate_tree_and_visualize/tree_plot"], "id": "#main/generate_tree_and_visualize"}], "outputs": [{"type": "File", "outputSource": "#main/calc_msa_and_dist_mat/dist_tsv", "id": "#main/dist_tsv"}, {"type": "File", "outputSource": "#main/generate_tree_and_visualize/tree_plot", "id": "#main/tree_plot"}], "id": "#main"}], "cwlVersion": "v1.0"}